### PR TITLE
Batching: DataStore + PredictFrontend

### PR DESCRIPTION
**Pull request by charlieyou**
Same as #375, but adding PredictFrontend to get CR comments on it.

PredictFrontend abstracts the predict callback so that Redis is bypassed when batching is not used.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: mms/predict_frontend.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+                logger.error(str(e))
+                abort(400, str(e))
+        else:
+            msg = '%s is not supported for input content-type' % input_type
+            if model_name + '_Prediction5XX' in MetricsManager.metrics:
+                MetricsManager.metrics[model_name + '_Prediction5XX'].update(metric=1)
+            logger.error(msg)
+            abort(500, "Service setting error. %s" % (msg))
+
+        if self.batching:
```
Comment (vdantu) [inline mms/predict_frontend.py]: Shouldn't this be in try/except block?

File: mms/predict_frontend.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                MetricsManager.metrics[model_name + '_Prediction5XX'].update(metric=1)
+            logger.error(msg)
+            abort(500, "Service setting error. %s" % msg)
+
+        logger.debug("Prediction request handling time is: %s ms",
+                     ((time.time() - handler_start_time) * 1000))
+        return self.handler.jsonify({'prediction': response})
+
+    def _batch_predict(self, input_data, input_type, output_type, model_name):
+        _id = str(uuid.uuid4())
```
Comment (vdantu) [inline mms/predict_frontend.py]: Does this ensure uniqueness? The responses from model-services are mapped to a particular frontend thread based on this right?

File: mms/serving_frontend.py [hunk 3]
```diff
@@ -40,10 +40,23 @@ def __init__(self, app_name):
...
+        def gen_prefix(prefix, length=5, chars=string.ascii_letters+string.digits):
+            return prefix + ''.join(random.choice(chars) for _ in range(length)) + '_'
+
         try:
             self.service_manager = ServiceManager()
             self.handler = FlaskRequestHandler(app_name)
 
+            if batching:
+                # generate a pseudo-random prefix to avoid collisions between runs
+                prefix = gen_prefix(app_name)
```
Comment (vdantu) [inline mms/serving_frontend.py]: Wouldn't this tie each front-end to its own model-services? You might want to consider removing this if you want other model-service threads to handle the queues.

File: docker/redis.conf [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+################################# GENERAL #####################################
+daemonize yes
+supervised no
+pidfile redis.pid
+loglevel notice
+logfile redis.log
+
+################################ SNAPSHOTTING  ################################
+save 900 1
```
Comment (vdantu) [inline docker/redis.conf]: Based on our discussion, disable this snapshotting for now. If the requirements exists to snapshot requests, we can re-do the changes.

File: mms/serving_frontend.py (hunk 3 above)
Comment (charlieyou) [reply mms/serving_frontend.py]: fixed

File: mms/predict_frontend.py (hunk 2 above)
Comment (charlieyou) [reply mms/predict_frontend.py]: Yes, here is a SO answer on uuid4: https://stackoverflow.com/questions/703035/when-are-you-truly-forced-to-use-uuid-as-part-of-the-design/786541#786541

"Frankly, in a single application space without malicious actors, the extinction of all life on earth will occur long before you have a collision, even on a version 4 UUID, even if you're generating quite a few UUIDs per second."

File: mms/predict_frontend.py (hunk 1 above)
Comment (charlieyou) [reply mms/predict_frontend.py]: done

File: mms/serving_frontend.py [hunk 5]
```diff
@@ -40,17 +38,14 @@ def __init__(self, app_name, batching=False, data_store_config=None):
...
-
         try:
             self.service_manager = ServiceManager()
             self.handler = FlaskRequestHandler(app_name)
 
             if batching:
-                # generate a pseudo-random prefix to avoid collisions between runs
-                prefix = gen_prefix(app_name)
-                data_store = DataStore(prefix, data_store_config)
+                # TODO generate a unique prefix in docker/mxnet-model-server and pass it into all gunicorn workers
```
Comment (vdantu) [inline mms/serving_frontend.py]: Good point to make a TODO. would this still be an issue after removing snapshotting?

**Pull request merged**